Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs

```diff
@@ -1,29 +1,648 @@
+                {
+                    new XRFeatureDescriptor()
+                    {
+                        name = "SimpleFeature",
+                        featureType = FeatureType.Binary
+                    }
+                }
+            }.ToJson()
+        }.ToJson());
+
+        InputSystem.Update();
+
+        var generatedLayout = InputSystem.TryLoadLayout("XRInput::Manufacturer::XRThisLayoutShouldhave1ValidName");
+        Assert.That(generatedLayout, Is.Not.EqualTo(null));
+    }
+
+    [Test]
+    [Category("Layouts")]
+    public void Layouts_XRLayoutFeaturesOnlyContainAllowedCharacters()
+    {
+        testRuntime.ReportNewInputDevice(
+            new InputDeviceDescription
+        {
+            interfaceName = XRUtilities.kXRInterface,
+            product = "XRDevice",
+            manufacturer = "__Manufacturer::",
+            capabilities = new XRDeviceDescriptor
+            {
+                deviceRole = DeviceRole.Generic,
+                inputFeatures = new List<XRFeatureDescriptor>()
+                {
+                    new XRFeatureDescriptor()
+                    {
+                        name = "SimpleFeature[|.:+-?<1",
+                        featureType = FeatureType.Binary
+                    }
+                }
+            }.ToJson()
+        }.ToJson());
+
+        InputSystem.Update();
+
+        var generatedLayout = InputSystem.TryLoadLayout("XRInput::Manufacturer::XRDevice");
+        Assert.That(generatedLayout, Is.Not.EqualTo(null));
+        Assert.That(generatedLayout.controls.Count, Is.EqualTo(1));
+
+        var childControl = generatedLayout.controls[0];
+        Assert.That(childControl.name, Is.EqualTo(new InternedString("SimpleFeature1")));
+    }
+
+    [Test]
+    [Category("Layouts")]
+    public void Layouts_XRDevicesWithNoOrInvalidCapabilitiesDoNotCreateLayouts()
+    {
+        testRuntime.ReportNewInputDevice(
+            new InputDeviceDescription
+        {
+            interfaceName = XRUtilities.kXRInterface,
+            product = "XRDevice1",
+            manufacturer = "XRManufacturer",
+            capabilities = null
+        }.ToJson());
+
+        testRuntime.ReportNewInputDevice(
+            new InputDeviceDescription
+        {
+            interfaceName = XRUtilities.kXRInterface,
+            product = "XRDevice2",
+            manufacturer = "XRManufacturer",
+            capabilities = "Not A JSON String"
+        }.ToJson());
+
+        InputSystem.Update();
+
+        var generatedLayout = InputSystem.TryLoadLayout("XRInput::XRManufacturer::XRDevice1");
+        Assert.That(generatedLayout, Is.EqualTo(null));
+
+        generatedLayout = InputSystem.TryLoadLayout("XRInput::XRManufacturer::XRDevice2");
+        Assert.That(generatedLayout, Is.EqualTo(null));
+    }
+
+    [Test]
+    [Category("Layouts")]
+    public void Layouts_WhenMatchingAKnownDeviceTheGeneratedLayoutInheritsProperly()
```

Comment:
Avoid "proper" and "correct" in test names and try to be specific instead. Also, this test seems to single out one case with no guarantee that the other cases work the same. Recommend converting this test into one that takes the product+manufacturer combination as well as the base layout name as a parameter and then uses `[TestCase(...)]`.

Reply:
Also recommend breaking up test names when there's a condition+outcome kinda setup. I.e. Devices_WhenThisIsTheCase_ThenThisIsTheResult. Just helps navigating test names a bit.